### PR TITLE
Lots of new features and refactoring

### PR DESCRIPTION
**Pull request by shahrukhqasim**
Including:

1. Change predict file so it uses a class so same code can be used in
   running full validation plots and for normal predictions.
2. Add code to do full validation after every few thousand batches it
   will run hyper param loop for 100 iterations and then produce full
   analysis plots so it is easy to monitor network performance.
3. Change analysis algorithm to graph matching and now writing graphs
   pickled instead of lists which are much easier to manage and there is
   no no need to predict and analyze manually for simple monitoring.

### CODE REVIEW

File: Train/june_format_example_nf_pca_double_coords_LCN.py [hunk 1]
```diff
@@ -16,6 +17,9 @@
 from tensorflow.keras import Model
 
 from experiment_database_reading_manager import ExperimentDatabaseReadingManager
+from hgcal_predictor import HGCalPredictor
+from hyperparam_optimzer import OCHyperParamOptimizer
```
Comment (jkiesele) [inline Train/june_format_example_nf_pca_double_coords_LCN.py:21]: let's fix the typo here and for the file itself (otherwise it'll be in there forever)

File: Train/june_format_example_nf_pca_double_coords_LCN.py [hunk 2]
```diff
@@ -255,6 +259,7 @@ def gravnet_model(Inputs,
 
 import training_base_hgcal
 train = training_base_hgcal.HGCalTraining(testrun=False, resumeSilently=True, renewtokens=False)
+train.val_data.writeToFile(train.outputDir + 'valsamples.djcdc')
```
Comment (jkiesele) [inline Train/june_format_example_nf_pca_double_coords_LCN.py:262]: why is this needed? this should be done automatically in principle

File: modules/GravNetLayersRagged.py [hunk 3]
```diff
@@ -2070,7 +2070,34 @@ def call(self, inputs):
         
         
         
-        
+class EyeInitializer(tf.keras.initializers.Initializer):
```
Comment (jkiesele) [inline modules/GravNetLayersRagged.py:2073]: Can we make a new file for initialisers? I think it's better than to put everything in layers.

File: modules/Layers.py [hunk 4]
```diff
@@ -586,7 +587,6 @@ def train_step(self, data):
         return ret_dict
 
 
-
-
 global_layers_list['ExtendedMetricsModel']=ExtendedMetricsModel
 global_layers_list['RobustModel']=RobustModel
+global_layers_list['EyeInitializer']=EyeInitializer
```
Comment (jkiesele) [inline modules/Layers.py:592]: This can remain as a workaround if needed

File: modules/hgcal_predictor.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+import os
+from keras.models import load_model
+from keras import backend as K
+from DeepJetCore.customObjects import get_custom_objects
+from DeepJetCore.training.gpuTools import DJCSetGPUs
+from DeepJetCore.training.training_base import custom_objects_list
+
+
+class HGCalPredictor():
```
Comment (jkiesele) [inline modules/hgcal_predictor.py:23]: should we change this upstream and clean up the prediction in DJC, so that you could inherit or something? This seems like a lot of code duplication here (but it's ok for now, just to keep in mind).

File: scripts/predict_hgcal.py [hunk 6]
```diff
@@ -25,128 +21,8 @@
...
-    td.clear()
-    gen.clear()
-
-    td.writeOutPredictionDict(dumping_data, args.outputDir + "/" + outfilename)
-    outputs.append(outfilename)
 
-with open(args.outputDir + "/outfiles.txt", "w") as f:
-    for l in outputs:
-        f.write(l + '\n')
+HGCalPredictor(args.inputSourceFileList, args.trainingDataCollection, args.outputDir, batchsize, unbuffered=False).predict(args.inputModel)
```
Comment (jkiesele) [inline scripts/predict_hgcal.py:25]: yeah that makes a lot of sense here (as I said above, maybe make the DJC version more flexible in the first place would be also an idea)

**Comment by jkiesele** [comment]
I've added a few comments.

File: Train/june_format_example_nf_pca_double_coords_LCN.py (hunk 2 above)
Comment (shahrukhqasim) [reply Train/june_format_example_nf_pca_double_coords_LCN.py:262]: It does it after a while. So I had to manually do it again since I needed to read the list of val files before.

File: modules/hgcal_predictor.py (hunk 5 above)
Comment (shahrukhqasim) [reply modules/hgcal_predictor.py:23]: it was already duplicated I just moved it to a class. At some point, we can change to be directly in DJC if you want. Right now, I don't wanna go there.
